NXP: Fix MCUX ADC12 instance defintions

henrikbrixandersen opened the pull request:
Switch the MCUX ADC12 driver from using DT_INST_* definitions to using DT_ADC_* dts fixup definitions in order to support configuration where e.g. only ADC_1 and ADC_2 are enabled.

henrikbrixandersen commented:
This change brings the MCUX ADC12 driver in line with how this is handled in the MCUX ADC16 driver.

galak reviewed (changes requested):
I'm not sure I understand why to move away from the `DT_INST` defines.  These are more generic as they don't need support in dts_fixup.h.  I'd rather see the ADC16 driver changed to use `DT_INST` if possible instead.

henrikbrixandersen commented:
> I'm not sure I understand why to move away from the `DT_INST` defines. These are more generic as they don't need support in dts_fixup.h. I'd rather see the ADC16 driver changed to use `DT_INST` if possible instead.

@galak I totally agree. I would like to move to the `DT_INST` defines all over. What we currently have in the ADC subsystem (and other subsystems) is a mix of the two. Currently the MCUX ADC12 driver does `#if CONFIG_ADC_0 /* handle DT_INST_0_NXP_KINETIS_ADC12 stuff */ #endif */` which causes compilation problems when e.g. having `CONFIG_ADC_0=n` (since the `DT_INST` numbering if now "off by one" compared to the `CONFIG_ADC_n` Kconfig variables).

Either we need to covert it to fully use the `DT_INST` defines (as in `#ifdef DT_INST_0_NXP_KINETIS_ADC12 /* handle DT_INST_0_NXP_KINETIS_ADC12  stuff */ #endif` or do what I did here. The latter has the advantage of adhering to the `CONFIG_ADC_n` Kconfig variables, thus making the samples dependent on these work out of the box.

What I would really like to do is to remove all the `CONFIG_ADC_n` Kconfig variables, convert all drivers to use `DT_INST` defines, and convert all the ADC samples/shell/... to work on DTS aliases. I believe this should be done across the tree, in other subsystems as well (I2C, SPI, ...).

henrikbrixandersen committed:
drivers: adc: mcux_adc12: use DT_INST defines for instances

Convert the NXP Kinetis ADC12 driver from relying on CONFIG_ADC_n
Kconfig defines to using DT_INST defines for instance configuration.

This resolves the issue of having e.g. ADC12 instances 2 and 3
enabled, but not instance 0.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
@galak I have updated the patch to use `DT_INST` defines for enabling the MCUX ADC12 driver instances. This also resolves the "off by one" issue.
